src/Rules.Framework/RulesEngine.cs

```diff
@@ -64,6 +64,39 @@ public Task<RuleOperationResult> AddRuleAsync(Rule<TContentType, TConditionType>
             return this.AddRuleInternalAsync(rule, ruleAddPriorityOption);
         }
 
+        /// <summary>
+        /// PGet the unique condition types associated with rules of a specific content type/>.
+        /// </summary>
+        /// <param name="contentType"></param>
+        /// <param name="dateBegin"></param>
+        /// <param name="dateEnd"></param>
+        /// <remarks>
+        /// <para>
+        /// A set of rules is requested to rules data source and all conditions are evaluated
+        /// against them to provide a set of matches.
+        /// </para>
+        /// <para>All rules matching supplied conditions are returned.</para>
+        /// </remarks>
+        /// <returns>the matched rule; otherwise, null.</returns>
+        public async Task<IEnumerable<IConditionNode<TConditionType>>> GetUniqueConditionTypesAsync(TContentType contentType, DateTime dateBegin, DateTime dateEnd)
+        {
+            var matchedRules = await this.rulesDataSource.GetRulesAsync(contentType, dateBegin, dateEnd).ConfigureAwait(false);
+
+            var conditionTypes = new List<IConditionNode<TConditionType>>();
+
+            if (!matchedRules.Any())
+            {
+                return conditionTypes;
+            }
+
+            foreach (var rule in matchedRules)
```

Comment:
This was not the point we wanted to achieve. The purpose was to return an `IEnumerable<TConditionType>` with all unique `TConditionType` values contained on `matchedRules`. This means, for each rule, visit the whole conditions tree and add to `conditionTypes` list as new `TConditionType` values are found.

Understand what I mean? We can get together for a few minutes if you want to discuss this 😄

Reply:
I've updated the code to separate the logic into a class outside of the RulesEngine, but I still don't understand what you mean by conditions tree, as a rules only has a root condition accessible.